### PR TITLE
Add convert CLI command to grafana-agent

### PR DESCRIPTION
**Pull request by erikbaranowski**
#### PR Description

most of this PR is about how we are setting up the CLI and the documentation in/around the CLI. I'm open to feedback and opinions from anyone on what this should look like after you take a look at the PR.

we will need some more extensive docs about this work in a new Migration guide section of our docs covered by https://github.com/grafana/agent/issues/4073

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: docs/sources/flow/reference/cli/convert.md
Comment (tpaschalis) [inline docs/sources/flow/reference/cli/convert.md:29]: Have we considered an `--output` flag instead? (and writing to stdout when it's not present)

IMHO overwriting the input configuration file might be confusing to users, cannot be retried, plus we have to keep track where the file came to check whether it can be used or not.

Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md]: I agree, updated

File: docs/sources/flow/reference/cli/convert.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/cli/convert.md:30]:
```suggestion
* `--source-format`, `-f`: Required. The format of the source file. Supported formats: 'prometheus'.
```

File: docs/sources/flow/reference/cli/convert.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/cli/convert.md:28]:
```suggestion
* `--output`, `-o`: The filepath and filename where the output is written.
```

**Review by clayton-cornell — APPROVED** [review]

**Review by clayton-cornell — COMMENTED** [review]
Docs are fine as-is. It matches the style/language of fmt. 

Future edit and rework could be done to clean up the language of all Agent CLI docs and clarify some things

File: docs/sources/flow/reference/cli/convert.md
Comment (tpaschalis) [inline docs/sources/flow/reference/cli/convert.md:21]: I don't think we need this restriction anymore with the `--output` flag, right?

File: cmd/internal/flowmode/cmd_convert.go
Comment (tpaschalis) [inline cmd/internal/flowmode/cmd_convert.go:64]: Is there an easy way to make this a required parameter? Otherwise it fails with the cryptic message
```
Error: (Error) unrecognized kind "" given to the config converter
```

**Review by tpaschalis — APPROVED** [review]
Just adding a couple of nits regarding ease of use.

Otherwise approving so I don't block things, I'm super happy we're having this!

File: cmd/internal/flowmode/cmd_convert.go
Comment (mattdurham) [inline cmd/internal/flowmode/cmd_convert.go:65]: Do we have documentation on what `warnings` mean?

File: converter/diag/diagnostics.go
Comment (mattdurham) [inline converter/diag/diagnostics.go:72]: nit: How does this behave on `cmd.exe`?

File: docs/sources/flow/reference/cli/convert.md
Comment (mattdurham) [inline docs/sources/flow/reference/cli/convert.md:24]: nit: Should both uses of file be changed to config to handle stdin/stdout

**Review by mattdurham — APPROVED** [review]
Minor nits but looks good

File: docs/sources/flow/reference/cli/convert.md
Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md:27]: Can you elaborate on this 1 or was it addressed with my last commit? I'm not sure what it is referring to in the context of this paragraph.

File: cmd/internal/flowmode/cmd_convert.go
Comment (rfratto) [inline cmd/internal/flowmode/cmd_convert.go:41]: You don't need an alias to the same command name, this is for alternative names

File: cmd/internal/flowmode/cmd_convert.go
Comment (rfratto) [inline cmd/internal/flowmode/cmd_convert.go:25]: If file is optional, then this should be:

```suggestion
		Use:   "convert [flags] [file]",
```

File: cmd/internal/flowmode/cmd_convert.go
Comment (rfratto) [inline cmd/internal/flowmode/cmd_convert.go:113]: I find it a bit surprising that the default behavior is that warnings are treated as errors, and you have to specifically bypass them.

IMO warnings should _not_ cause the conversion to fail, but you should instead allow a flag to fail on warnings.

File: cmd/internal/flowmode/cmd_convert.go
Comment (rfratto) [inline cmd/internal/flowmode/cmd_convert.go:126]: nit: use `os.Create`, which will give you the same effect as this.

File: converter/diag/diagnostics.go
Comment (rfratto) [inline converter/diag/diagnostics.go:76]: nit: use `strings.Builder` instead, which is more efficient and also handles newlines for you:

```suggestion
	var sb strings.Builder
	for ix, diag := range ds {
		fmt.Fprint(&sb, diag.Error())
		if ix+1 < len(ds) {
			fmt.Fprintln(&sb) 
		}
	}

	return sb.String()
```

File: docs/sources/flow/reference/cli/convert.md
Comment (rfratto) [inline docs/sources/flow/reference/cli/convert.md:9]: River is a proper noun:

```suggestion
to a Grafana Agent Flow River file.
```

File: docs/sources/flow/reference/cli/convert.md
Comment (rfratto) [inline docs/sources/flow/reference/cli/convert.md:27]:
```suggestion
configuration or cannot be converted to a Grafana Agent Flow River file.
```

File: docs/sources/flow/reference/cli/convert.md
Comment (rfratto) [inline docs/sources/flow/reference/cli/convert.md:33]: I think we're burying the lede here with a lack of explanation for what `prometheus` here means. 

I'd like to see a section for supported source formats, which go into a little more detail about them. For example, we should document:

* What version of Prometheus is used for conversion, potentially even linking to its documentation
* What broad features are/aren't supported (recording rules, alerting rules, etc)
* Anything else you might think is useful to know about the Prometheus conversion

**Review by rfratto — APPROVED** [review]
Nice work! I'm excited to start playing around with this.

File: cmd/internal/flowmode/cmd_convert.go
Comment (erikbaranowski) [reply cmd/internal/flowmode/cmd_convert.go:112]: I think it's a fair point on the semantics. That being said, I think it's important that our default behavior for running the agent is not to start up automatically if we have unsupported features in the conversion and I was making this behave the same way for consistency.

I'm open to suggestion here if we can change the semantics in a way that makes sense and gives us the behavior we want. Do you agree on the behavior when running should not start without an explicit flag if something in the source file is unsupported?

File: docs/sources/flow/reference/cli/convert.md
Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md:33]: @clayton-cornell what do you think about the best way to organize this information? I was anticipating most of these details would exist in the migration guide for https://github.com/grafana/agent/issues/4073 . If we want to include that work as part of this PR before merge I think that's also fair.

I don't have a strong opinion about the location of this information, but I do agree we want it before I consider the converter 1.0 'done' end to end.

File: cmd/internal/flowmode/cmd_convert.go
Comment (erikbaranowski) [reply cmd/internal/flowmode/cmd_convert.go:112]: An alternative brainstorm is to have them also be errors but introduce an error type to the Diagnostic so we can differentiate and add the right semantic flag for ignoring them.

File: docs/sources/flow/reference/cli/convert.md
Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md:33]: Proposal: do we want to make this the staging branch and do the full docs/run CLI changes to this branch then merge it to main all together?

Let's still coalesce on the convert CLI implementation on this PR as written.

File: cmd/internal/flowmode/cmd_convert.go
Comment (rfratto) [reply cmd/internal/flowmode/cmd_convert.go:112]: After talking about this offline, I think we need more time to explore what types of warnings we might generate, and if all of them fall into the same category of needing investigation or being OK for the user to ignore. 

For example, are the warnings for these two the same severity: 

1. Alerting is configured in Prometheus but no equivalent alerting config was generated in Flow
2. The wal_directory argument in static mode is ignored because the WAL storage happens at a higher level in Flow 

To me, the first one seems more fatal than the second, while the second one is more of a actual heads-up style of warning. 

We'll be able to classify these better after we're done with writing all the converters, so my suggestion for now would be to leave the behavior as-is and revisit changing whether warnings are fatal later, since making them non-fatal would be a backwards compatible change, but the inverse wouldn't.

File: docs/sources/flow/reference/cli/convert.md
Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md:33]: I'm going to draft some additional docs for this PR and let's go from there

File: docs/sources/flow/reference/cli/convert.md
Comment (mattdurham) [reply docs/sources/flow/reference/cli/convert.md]: IE the input can be stdin and output can be stdout and not just files. 

```
The command fails if the config being converted has syntactically incorrect
configuration or cannot be converted to a Grafana Agent Flow river config.
```

Its a super minor opinion and if you think file is clearer keep it.

Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md]: got it now, thanks for explaining. I'm updating throughout for my next pass with extra docs.

File: docs/sources/flow/reference/cli/convert.md
Comment (clayton-cornell) [reply docs/sources/flow/reference/cli/convert.md:33]: @erikbaranowski This is something we've chatted a little about in an offline 1:1 and in messages and it's in-0line with my earlier concerns about this not really providing all info and being confusing to me when I read it. I opted to say "it's follwoing the `fmt` doc, so... I'll leave it as-is for now. It could definitely have a rethink - and if we do document this a bit differently, we should revisit (in another PR?) the `fmt` doc... and maybe `run` as well.

For here, maybe this: follow manpage style for the CLI documentation. Stick to the basics... this is the command and syntax, and here are the arguments you can use... (these are required, these are optional) followed by a common usecase example.

Then for more details on implementation and so on, we shift that to a migration guide that explains the "why woudl I use this" in more detail.

For where to include... keep the PRs simple.  We can keep this for code changes and the basic CLI doc content.  This makes reviews a bit more manageable (bite sized).

Comment (erikbaranowski) [reply docs/sources/flow/reference/cli/convert.md]: thanks all, @clayton-cornell can you take 1 more pass on the last commit and see if you want to tune anything for now. I appreciate all the feedback and hope to keep iterating on this.

File: docs/sources/flow/reference/cli/convert.md
Comment (ptodev) [inline docs/sources/flow/reference/cli/convert.md:19]: Not sure why we list `--output` and `--bypass-warnings` both here and in the list below? I think there's no need to list them here, we could just have them in the list.

File: docs/sources/flow/reference/cli/convert.md
Comment (ptodev) [inline docs/sources/flow/reference/cli/convert.md:51]: > Warnings are defined as non-blocking issues identified during the conversion
for feature[s] on the source file not supported in Grafana Agent Flow.

If the source config uses a feature which is not available in Flow, then wouldn't that have to be an error? If the feature is missing from the Flow config, wouldn't that mean it has reduced functionality?

File: cmd/internal/flowmode/cmd_convert.go
Comment (ptodev) [reply cmd/internal/flowmode/cmd_convert.go:112]: It might also help to have different categories of errors and warnings and document them. Or we could have different error codes for each type of error/warning. Then the customers could choose which warnings/errors to enable or disable. It'd also make the docs more clear about what kinds of issues could be reported.

**Review by ptodev — APPROVED** [review]
I left a few minor comments, but I'm happy for the PR to be merged even if they're not addressed.

**Pull request merged**
